Add omp_for_schedule and omp_for_nowait_schedule policies

willkill07 opened the pull request:
# Summary

- This PR is a feature
- It does the following (modify list as needed):
  - Modifies/refactors OpenMP policies for CPU by augmenting schedule types
  - Adds Guided, Runtime, Auto, and Dynamic scheduling options per the discussion at the last RAJA dev meeting

# Design review

This introduces two new end-user policies: `omp_for_schedule_exec<T>` and `omp_for_nowait_schedule_exec<T>` where:
 -  `T` is a scheduling policy
 - `omp_for_schedule_exec` only specifies the scheduling policy
 - `omp_for_nowait_schedule_exec` also includes the nowait clause on the loop

A scheduling policy is one of:
- `RAJA::policy::omp::Dynamic<N>`
- `RAJA::policy::omp::Static<N>`
- `RAJA::policy::omp::Guided<N>`
- `RAJA::policy::omp::Auto`
- `RAJA::policy::omp::Runtime`

This uses tag dispatch in an internal namespace to `RAJA::policy::omp` to reduce lookup.

Any feedback is welcome.

# TODO
- [x] Handle compiler workarounds for MSVC
- [x] Write unit/functional tests for each policy
- [x] Add RAJA::policy::omp schedule clauses to documentation

willkill07 commented:
@rhornung67 and @trws -- any feedback here would be good. I'm flexible in terms of any reasonable solution. I know this won't scale with a product of options.

trws commented:
I need to think about this a bit, but am fine with the concept.  First blush I’m thinking we can collapse this a little bit so we don’t have quite such a profusion without losing much if we can figure out how to spell it.  Reasoning:

* auto is what you get when you don’t use a schedule clause
* static is unique and needs to stand alone, could have dynamic or static chunk size
* dynamic, guided and runtime tend to all be implemented by the same fully dynamic code path, so we could cover all three of these and static with a dynamic chunk size with a single impl that uses `schedule(runtime, chunk)` with `omp_set_schedule`

The no wait is actually where I’m wondering what we should do.  I keep thinking we’d do well to have some expected semantics from Forall, but we really kinda don’t, it’s just defined as “whatever the policy does in the underlying model” and that may become hard to reason about as the async kernels stuff starts to get in heavier use.

For what it’s worth, I’d be fine with taking this as is if it’s a use-case we want to support.  It just has me thinking about how much of the OpenMP interface we want to surface through raja to the user, especially where it’s only really usable inside a region.

willkill07 commented:
@trws -- thanks

I agree with auto and the importance of static. I'm also viewing this as a testbed for any directive-based programming model where we have various directives and clauses.

The single code path for dynamic / guided with `omp_set_schedule` may incur overhead in target offload kernels. I'm not sure what the best approach is.

NVIDIA's new nvc++ compiler looks to be a promising candidate for modern C++ and directive-based abstractions with OpenACC.

MrBurmark commented on include/RAJA/policy/openmp/policy.hpp:
```diff
@@ -76,12 +118,7 @@ struct omp_for_nowait_exec
-                                                              Platform::host,
-                                                              omp::For,
-                                                              omp::Static<N>> {
+struct omp_for_static : omp_for_schedule<omp::Static<N>> {
```
Could this be an alias to avoid generating another class?

rhornung67 replied:
You mean the omp::Static part?  Will was working on a PR to clean up some of this. We briefly discussed at the RAJA meeting earlier this week. I'll defer to the OpenMP experts.

willkill07 commented on include/RAJA/policy/openmp/forall.hpp:
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
+  /// Tag dispatch for omp forall
+
+  template <typename Iterable, typename Func, unsigned int ChunkSize>
+  RAJA_INLINE void forall_impl(const ::RAJA::policy::omp::Auto&,
```
auto is our "old" default

willkill07 commented on include/RAJA/policy/openmp/forall.hpp:
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
-/// OpenMP parallel for policy implementation
-///
+  template <typename Iterable, typename Func, unsigned int ChunkSize>
+  RAJA_INLINE void forall_impl(const ::RAJA::policy::omp::Static<ChunkSize>&,
```
explicit specialization for Static<ChunkSize>

willkill07 commented on include/RAJA/policy/openmp/forall.hpp:
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
-template <typename Iterable, typename Func>
-RAJA_INLINE void forall_impl(const omp_for_exec&,
+  template <typename Iterable, typename Func>
+  RAJA_INLINE void forall_impl(const ::RAJA::policy::omp::Runtime&,
```
explicit specialization for Runtime

willkill07 commented on include/RAJA/policy/openmp/forall.hpp:
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
+  }
+
+  // dynamic & guided
+  template <typename Policy, typename Iterable, typename Func, unsigned int ChunkSize>
```
template argument deduction for runtime and guided -- using `omp_set_schedule` to choose what to do

willkill07 commented on include/RAJA/policy/openmp/forall.hpp:
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
-  for (decltype(distance_it) i = 0; i < distance_it; ++i) {
-    loop_body(begin_it[i]);
-  }
+  internal::forall_impl(Schedule{}, std::forward<Iterable>(iter), std::forward<Func>(loop_body));
```
internal tagged dispatch to schedule specified

willkill07 commented on include/RAJA/policy/openmp/policy.hpp:
```diff
@@ -22,14 +22,32 @@
+{
+    template <omp_sched_t Sched, int Chunk>
+    struct Schedule {
+        constexpr static omp_sched_t schedule = Sched;
```
`schedule` will be negative if `Auto`
`chunk_size` will be negative if "default"

willkill07 commented on include/RAJA/policy/openmp/policy.hpp:
```diff
@@ -42,10 +60,23 @@ struct For {
 
-template <unsigned int ChunkSize>
-struct Static : std::integral_constant<unsigned int, ChunkSize> {
+static constexpr int default_chunk_size = -1;
```
This is a *public* interface option for users. The OpenMP runtime says that a "default" will be used when a passed value into `omp_set_schedule` is negative

willkill07 commented on include/RAJA/policy/openmp/policy.hpp:
```diff
@@ -42,10 +60,23 @@ struct For {
-struct Static : std::integral_constant<unsigned int, ChunkSize> {
+static constexpr int default_chunk_size = -1;
+
+struct Auto : private internal::Schedule<omp_sched_auto, default_chunk_size>{
```
`Auto`, `Static`, and `Runtime` are new types so tagged dispatch works correctly

willkill07 commented on include/RAJA/policy/openmp/policy.hpp:
```diff
@@ -55,52 +86,58 @@ struct Static : std::integral_constant<unsigned int, ChunkSize> {
 //////////////////////////////////////////////////////////////////////
 //
 
+namespace exec {
```
I placed all execution policy names in the exec namespace so users can autocomplete `RAJA::policy::omp::` to get the various schedules and `default_chunk_size`

willkill07 commented on include/RAJA/policy/openmp/policy.hpp:
```diff
@@ -55,52 +86,58 @@ struct Static : std::integral_constant<unsigned int, ChunkSize> {
+                                                              omp::For,
+                                                              omp::NoWait,
+                                                              Schedule> {
+    static_assert(std::is_base_of<::RAJA::policy::omp::internal::Schedule, Schedule>::value,
```
Compile-time error for users who don't specify a correct schedule

willkill07 commented on include/RAJA/policy/openmp/policy.hpp:
```diff
@@ -55,52 +86,58 @@ struct Static : std::integral_constant<unsigned int, ChunkSize> {
+        "Schedule must be one of: Auto|Runtime|Static|Dynamic|Guided");
 };
 
+using omp_for_exec = omp_for_schedule<Auto>;
```
All other policies now alias to `omp_for_schedule` or `omp_for_nowait_schedule`

willkill07 commented on include/RAJA/policy/openmp/forall.hpp:
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
+  {
+    omp_sched_t prev_sched;
+    int prev_chunk;
+    omp_get_schedule(&prev_sched, &prev_chunk);
```
@trws is preserving the old schedule settings important? I have this added in and don't think it should add too much to the implementation (only used for `Guided` and `Dynamic`)

trws replied:
As long as it only exists for guided and dynamic I have no issues with it at all.  The dynamic code paths tend to have a good deal more overhead for startup anyway, and are less frequently used.  There may be an implementation at some point where we decide to care, but none exists today that I know of.

willkill07 commented:
The HIP CI tests reveal something not-so-great:

```
In file included from /home/raja/include/RAJA/pattern/kernel/Region.hpp:23:
938/home/raja/include/RAJA/policy/openmp/policy.hpp:25:10: fatal error: 'omp.h' file not found
939#include <omp.h>
```

Is there a reason why `Region.hpp` is even including `openmp/policy.hpp` when nothing explicitly depends on it?

willkill07 commented:
Additionally, MSVC only supports OpenMP 2.0 -- the `omp.h` functions and types:
* `omp_set_schedule`
* `omp_get_schedule`
* `omp_sched_t`

Only were added with OpenMP 3.0. Should I add an `#if _OPENMP > 200805` guard and limit the code path for what's visible?

rhornung67 commented:
> Additionally, MSVC only supports OpenMP 2.0 -- the `omp.h` functions and types:
> 
>     * `omp_set_schedule`
> 
>     * `omp_get_schedule`
> 
>     * `omp_sched_t`
> 
> 
> Only were added with OpenMP 3.0. Should I add an `#if _OPENMP > 200805` guard and limit the code path for what's visible?

I think that adding the guard makes sense.  However, it seems odd that the compiler does not support OpenMP 3.0

rhornung67 commented:
> The HIP CI tests reveal something not-so-great:
> 
> ```
> In file included from /home/raja/include/RAJA/pattern/kernel/Region.hpp:23:
> 938/home/raja/include/RAJA/policy/openmp/policy.hpp:25:10: fatal error: 'omp.h' file not found
> 939#include <omp.h>
> ```
> 
> Is there a reason why `Region.hpp` is even including `openmp/policy.hpp` when nothing explicitly depends on it?

Probably an oversight after some code regorg. If that header is not needed, please remove its inclusion. Hopefully, it won't result in a cascade of other files relying on it being included here.

willkill07 commented:
@rhornung67 I made those changes and have a clean build across.

**Documentation**: Added, but a build wasn't triggered on readthedocs. I manually inspected a local build, though
**Testing**: Added to forall and atomic testing, but it's only enabled with `RAJA_TEST_EXHAUSTIVE`

@trws I think changes are at a point for a review. Not sure who else would be interested in reviewing this.

willkill07 commented:
> I'm pretty happy. @willkill07, do you have time to update to match main?

@trws Wouldn't this go into `develop`? -- or should change the target merge to `main`?

rhornung67 commented:
@willkill07 please merge into develop as your PR is set up to do.

willkill07 commented:
@rhornung67 this can be merged once checks pass (I cannot merge)

rhornung67 commented:
Will do.

rhornung67 commented:
@willkill07 merging your OpenMP policy PR.
